openid/openid_test.go

```diff
@@ -0,0 +1,295 @@
+	server:  usso.ProductionUbuntuSSOServer,
+	verifyF: verifySuccess,
+	expectResponse: &openid.Response{
+		ID: "https://login.ubuntu.com/+id/AAAAAA",
+	},
+}, {
+	about:       "bad url",
+	url:         "://return.to",
+	server:      usso.ProductionUbuntuSSOServer,
+	verifyF:     verifySuccess,
+	expectError: "parse ://return.to: missing protocol scheme",
+}, {
+	about:       "unexpected OP",
+	url:         "http://return.to?openid.ns=http://specs.openid.net/auth/2.0&openid.mode=id_res&openid.op_endpoint=https://login.ubuntu.com/%2Bopenid&openid.claimed_id=https://login.ubuntu.com/%2Bid/AAAAAA&openid.identity=https://login.ubuntu.com/%2Bid/AAAAAA&openid.return_to=http://return.to&openid.response_nonce=2005-05-15T17:11:51ZUNIQUE&openid.assoc_handle=1&openid.signed=op_endpoint,return_to,response_nonce,assoc_handle,claimed_id,identity&openid.sig=AAAA",
+	server:      usso.StagingUbuntuSSOServer,
+	verifyF:     verifySuccess,
+	expectError: `OpenID response from unexpected endpoint "https://login.ubuntu.com/\+openid"`,
+}, {
+	about:  "verification failure",
+	url:    "http://return.to?openid.ns=http://specs.openid.net/auth/2.0&openid.mode=id_res&openid.op_endpoint=https://login.ubuntu.com/%2Bopenid&openid.claimed_id=https://login.ubuntu.com/%2Bid/AAAAAA&openid.identity=https://login.ubuntu.com/%2Bid/AAAAAA&openid.return_to=http://return.to&openid.response_nonce=2005-05-15T17:11:51ZUNIQUE&openid.assoc_handle=1&openid.signed=op_endpoint,return_to,response_nonce,assoc_handle,claimed_id,identity&openid.sig=AAAA",
+	server: usso.ProductionUbuntuSSOServer,
+	verifyF: func(*gc.C) func(string, yopenid.DiscoveryCache, yopenid.NonceStore) (string, error) {
+		return func(string, yopenid.DiscoveryCache, yopenid.NonceStore) (string, error) {
+			return "", errors.New("TEST!")
+		}
+	},
+	expectError: `TEST!`,
+}, {
+	about:  "uses default NonceStore",
+	url:    "http://return.to?openid.ns=http://specs.openid.net/auth/2.0&openid.mode=id_res&openid.op_endpoint=https://login.ubuntu.com/%2Bopenid&openid.claimed_id=https://login.ubuntu.com/%2Bid/AAAAAA&openid.identity=https://login.ubuntu.com/%2Bid/AAAAAA&openid.return_to=http://return.to&openid.response_nonce=2005-05-15T17:11:51ZUNIQUE&openid.assoc_handle=1&openid.signed=op_endpoint,return_to,response_nonce,assoc_handle,claimed_id,identity&openid.sig=AAAA",
+	server: usso.ProductionUbuntuSSOServer,
+	verifyF: func(c *gc.C) func(string, yopenid.DiscoveryCache, yopenid.NonceStore) (string, error) {
```

Comment:
This could be simpler I think. if you defined verifyF as func(*gc.C, string, yopenid.DiscoveryCache, yopenid.NonceStore) and put the function constructor inside the test code and then did:

*openid.VerifyFunc = func(s string, c yopenid.DiscoveryCache, n yopenid.NonceStore) {
     test.verifyF(c, s, c, n)
}

Then you wouldn't need the nesting in every test case.